File: test/barcode_bridge_test.exs
# test/barcode_bridge_test.exs
#
# CodeRegistry grades gtin/acl13/cip13 as :barcode — "a merge bridged SOLELY by one of these is
# SUSPECT" — but suspicion is not the same as refusal. Most real matches are barcode matches, and
# a listing that carries nothing but a barcode has no other way to attach to anything.
#
# So the rule is narrower than "hold every barcode bridge": hold only when BOTH sides already
# stand alone — each carries a non-barcode code of its own — and the reassignable code is the only
# thing tying them together. That is the shape of an NHSBSA GTIN transfer, and nothing else.

defmodule BarcodeBridgeTest do
  use ExUnit.Case, async: true
  import Substrate, only: [claim: 5]

  @d ~D[2026-01-05]
  @gtin {:gtin, "05012345678900"}

  describe "a barcode still bridges when one side has nothing else" do
    test "a bare barcode listing attaches to a fully identified product" do
      # Shop B knows only the barcode. Refusing the bridge would orphan it for no gain.
      clusters =
        variants([
          listing("A", [{:cnk, "1111111"}, @gtin]),
          listing("B", [@gtin])
        ])

      assert length(clusters) == 1
    end

    test "two bare barcode listings are one product" do
      clusters = variants([listing("A", [@gtin]), listing("B", [@gtin])])

      assert length(clusters) == 1
    end

    test "a shared non-barcode code still bridges normally" do
      clusters =
        variants([
          listing("A", [{:cnk, "1111111"}, {:supplier_ref, "S-9"}]),
          listing("B", [{:cnk, "1111111"}, {:supplier_ref, "S-4"}])
        ])

      assert length(clusters) == 1
    end
  end

  describe "a barcode does not bridge two things that already stand alone" do
    test "two packs with their own supplier references are held apart" do
      # The NHSBSA case: one barcode, two dm+d packs, each with its own reference and no
      # national code anywhere. Today this fused silently.
      clusters =
        variants([
          listing("100g", [@gtin, {:supplier_ref, "AMPP-100g"}]),
          listing("30g", [@gtin, {:supplier_ref, "AMPP-30g"}])
        ])

      assert length(clusters) == 2
    end

    test "the held barcode is reported as a conflict, not silently dropped" do
      events =
        decide([
          listing("100g", [@gtin, {:supplier_ref, "AMPP-100g"}]),
          listing("30g", [@gtin, {:supplier_ref, "AMPP-30g"}])
        ])

      assert Enum.any?(events, &match?(%Events.ConflictFlagged{subject: {:identity_conflict, @gtin}}, &1))
      assert Enum.any?(events, &match?(%Events.ConflictFlagged{subject: {:merge, _}}, &1))
    end

    test "the hold does not depend on the order the listings arrive in" do
      claims = [
        listing("100g", [@gtin, {:supplier_ref, "AMPP-100g"}]),
        listing("30g", [@gtin, {:supplier_ref, "AMPP-30g"}])
      ]

      assert variants(claims) == variants(Enum.reverse(claims))
      assert length(variants(claims)) == 2
    end

    test "trusted same evidence overrides the hold" do
      clusters =
        variants([
          listing("100g", [@gtin, {:supplier_ref, "AMPP-100g"}]),
          listing("30g", [@gtin, {:supplier_ref, "AMPP-30g"}]),
          evidence(:same, {:supplier_ref, "AMPP-100g"}, {:supplier_ref, "AMPP-30g"})
        ])

      assert length(clusters) == 1
    end
  end

  describe "the rule leaves the existing national-code guard alone" do
    test "two different national codes on one barcode are still held" do
      clusters =
        variants([
          listing("A", [{:cnk, "1111111"}, @gtin]),
          listing("B", [{:cnk, "2222222"}, @gtin])
        ])

      assert length(clusters) == 2
    end

    test "a non-barcode code bridging two stand-alone sides still fuses" do
      # Only barcodes are suspect. A shared CNK joining two supplier listings is exactly the
      # match the system exists to make.
      clusters =
        variants([
          listing("A", [{:cnk, "1111111"}, {:supplier_ref, "S-9"}]),
          listing("B", [{:cnk, "1111111"}, {:supplier_ref, "S-4"}, @gtin])
        ])

      assert length(clusters) == 1
    end
  end

  # ── helpers ─────────────────────────────────────────────────────────────────

  defp listing(ref, codes), do: claim(:supplier, :identity, %{ref: ref, codes: codes}, @d, @d)

  defp evidence(relation, left, right),
    do: claim(:steward, :identity_evidence, %{relation: relation, left: left, right: right}, @d, @d)

  defp variants(claims), do: claims |> stamp() |> Substrate.current() |> Cluster.variants()

  defp decide(claims) do
    IdentityLedger.decide(IdentityLedger.new(), {:reconcile, variants(claims), @d})
  end

  defp stamp(claims), do: Enum.map(Enum.with_index(claims, 1), fn {c, i} -> %{c | order: i} end)
end

File: php/tests/BarcodeBridgeTest.php
<?php

declare(strict_types=1);

namespace Ingot\Tests;

use Ingot\Claim;
use Ingot\Cluster;
use Ingot\ConflictFlagged;
use Ingot\Events;
use Ingot\IdentityLedger;
use Ingot\Substrate;
use PHPUnit\Framework\TestCase;

/**
 * PHP parity for test/barcode_bridge_test.exs.
 *
 * A barcode is reassignable, so it must not be the one thing fusing two components that each
 * already carry an identifier of their own. When either side has nothing but barcodes, the join
 * still stands — refusing it would orphan a listing for no gain.
 */
final class BarcodeBridgeTest extends TestCase
{
    private const GTIN = ['gtin', '05012345678900'];

    public function test_a_bare_barcode_listing_attaches_to_an_identified_product(): void
    {
        $clusters = $this->variants([
            $this->listing('A', [['cnk', '1111111'], self::GTIN]),
            $this->listing('B', [self::GTIN]),
        ]);

        self::assertCount(1, $clusters);
    }

    public function test_two_bare_barcode_listings_are_one_product(): void
    {
        $clusters = $this->variants([
            $this->listing('A', [self::GTIN]),
            $this->listing('B', [self::GTIN]),
        ]);

        self::assertCount(1, $clusters);
    }

    public function test_two_packs_with_their_own_references_are_held_apart(): void
    {
        $clusters = $this->variants([
            $this->listing('100g', [self::GTIN, ['supplier_ref', 'AMPP-100g']]),
            $this->listing('30g', [self::GTIN, ['supplier_ref', 'AMPP-30g']]),
        ]);

        self::assertCount(2, $clusters);
    }

    public function test_the_held_barcode_is_reported_as_a_conflict(): void
    {
        $events = IdentityLedger::decide(IdentityLedger::new(), [
            'reconcile',
            $this->variants([
                $this->listing('100g', [self::GTIN, ['supplier_ref', 'AMPP-100g']]),
                $this->listing('30g', [self::GTIN, ['supplier_ref', 'AMPP-30g']]),
            ]),
            'd1',
        ]);

        $flagged = false;
        foreach ($events as $event) {
            if ($event instanceof ConflictFlagged && ($event->subject[0] ?? null) === 'identity_conflict') {
                $flagged = true;
            }
        }
        self::assertTrue($flagged, 'expected the held barcode to be flagged');
    }

    public function test_the_hold_does_not_depend_on_listing_order(): void
    {
        $claims = [
            $this->listing('100g', [self::GTIN, ['supplier_ref', 'AMPP-100g']]),
            $this->listing('30g', [self::GTIN, ['supplier_ref', 'AMPP-30g']]),
        ];

        self::assertCount(2, $this->variants($claims));
        self::assertEquals($this->variants($claims), $this->variants(array_reverse($claims)));
    }

    public function test_trusted_same_evidence_overrides_the_hold(): void
    {
        $clusters = $this->variants([
            $this->listing('100g', [self::GTIN, ['supplier_ref', 'AMPP-100g']]),
            $this->listing('30g', [self::GTIN, ['supplier_ref', 'AMPP-30g']]),
            Substrate::claim('steward', 'identity_evidence', [
                'relation' => 'same',
                'left' => ['supplier_ref', 'AMPP-100g'],
                'right' => ['supplier_ref', 'AMPP-30g'],
            ], 'd1', 'd1'),
        ]);

        self::assertCount(1, $clusters);
    }

    public function test_a_non_barcode_bridge_between_stand_alone_sides_still_fuses(): void
    {
        $clusters = $this->variants([
            $this->listing('A', [['cnk', '1111111'], ['supplier_ref', 'S-9']]),
            $this->listing('B', [['cnk', '1111111'], ['supplier_ref', 'S-4'], self::GTIN]),
        ]);

        self::assertCount(1, $clusters);
    }

    public function test_the_national_code_guard_is_unchanged(): void
    {
        $clusters = $this->variants([
            $this->listing('A', [['cnk', '1111111'], self::GTIN]),
            $this->listing('B', [['cnk', '2222222'], self::GTIN]),
        ]);

        self::assertCount(2, $clusters);
    }

    // ── helpers ──────────────────────────────────────────────────────────────

    /** @param list<array{0: string, 1: string}> $codes */
    private function listing(string $ref, array $codes): Claim
    {
        return Substrate::claim('supplier', 'identity', ['ref' => $ref, 'codes' => $codes], 'd1', 'd1');
    }

    /**
     * @param list<Claim> $claims
     *
     * @return list<array<string, array{0: string, 1: string}>>
     */
    private function variants(array $claims): array
    {
        $stamped = [];
        $i = 1;
        foreach ($claims as $claim) {
            $stamped[] = $claim->withOrder($i++);
        }

        return Cluster::variants(Substrate::current($stamped));
    }
}
